### ISSUE TITLE
Unable to successfully run macrobenchmark test based off on sample

### ISSUE THREAD

**Issue by spinEdamageD** [android/performance-samples#88]
I'm trying to run a simple startup test similar to the one shown in the sample here, and although I seem to have everything as shown in the sample gradle file, I keep getting an error when running the test. I'm trying to run directly on a FB device through the gcloud command, using the same command shown in the sample, and although the test seems to start correctly, I keep seeing the following:

`java.lang.NoClassDefFoundError: Failed resolution of: Lkotlin/collections/CollectionsKt;
at com.sample.macrotests.ColdStartupTest.startup(ColdStartupTest.kt:23)
... 30 trimmed
Caused by: java.lang.ClassNotFoundException: kotlin.collections.CollectionsKt
... 32 more`

this points to the line where I'm just setting the type of test to run, like:

`metrics = listOf(StartupTimingMetric()),`

I seem to be missing a dependency but I have everything as shown in the sample so I haven't been able to figure out exactly what I'm missing so not sure if it's a real issue or just something on my side I may be missing so just wanted to call it out here.

Thanks!

**Comment by ChrisCraik** [comment]
That class is part of the kotlin standard library, which should be bundled into your macrobenchmark apk via the library's dependency.

1) Are you building with gradle, 7.0 or higher?
2) Are you setting `experimentalProperties["android.experimental.self-instrumenting"] = true` in your macrobenchmark build.gradle as is done [here](https://github.com/android/performance-samples/blob/80c78f369fb5ee17ba45efd6dd4102b2c2518310/MacrobenchmarkSample/macrobenchmark/build.gradle#L32)? That's changes the build structure significantly, and allows the test to compile completely separately from the app.
3) Are you using minification in your macrobenchmark module build? That shouldn't be necessary, and won't affect your measurements, since the macrobenchmark builds separately from the app.

**Comment by spinEdamageD** [comment]
Hey Chris, thanks for the comment.

As for your questions:

1. Yeah, I'm using version 7.0.1
2. I actually have it as `properties["android.experimental.self-instrumenting"] = true` because if I have it as shown in the docs, I get the following error: `Could not get unknown property 'experimentalProperties' for extension 'android' of type com.android.build.gradle.TestExtension`. I don't know if this could be an issue but I'm pretty sure I saw that on another working sample so I don't think that's the reason behind the failure.
3. I actually am creating a minimized build for both main and test apk as if I don't add the minimization into both, I get a compilation error complaining about this.

Do you think this might be coming from point 2?

**Comment by ChrisCraik** [comment]
If it can't find the property, it's likely you need to update your android gradle plugin version. The sample uses 7.0.0-beta04, that or newer should work fine.

**Comment by spinEdamageD** [comment]
Hey man. Yeah, I've been trying to update that, but for my particular project that seems to be a bit challenging. 

Do you know for a fact that it won't work like `properties["android.experimental.self-instrumenting"] = true` ? I've been trying to find exactly where but I could've sworn I saw it like that in another working sample.

**Comment by ChrisCraik** [comment]
Without an update to a newer version of the AGP (android gradle plugin), you can make it work, but you won't get the full functionality of the studio / agp integration. You'll also have to have a different module structure temporarily.

For more information, see https://issuetracker.google.com/issues/193896397#comment2

That thread links to an older version of the docs, and the sample, which was built before the AGP support for macrobenchmarks.

**Comment by spinEdamageD** [comment]
Hey man, thanks again for the quick response. I took a look at the link you mentioned, and it looks like it should work with an older AGP version, because actually in my case I don't even want to run this via AS and I don't really care about trace files but rather I would like to integrate it into a CI flow and I just really care about the single results json file (as of now I just want to be able to trigger the run directly in FB using the `gcloud` command as seen in the sample), so from the comment in the link, it should work as it is with an older AGP version, so I'm not entirely sure this particular failure is because of this. 

I'm still trying to update the AGP version, but I'm out of other ideas as to what to try to make this work.

**Comment by ChrisCraik** [comment]
If you just care about CI results, you're totally fine going with the `com.android.library` approach documented in that bug. Basically you have to choose one of the two build structures:

1) `com.android.library`
Works on older versions of gradle (back to at least AGP 4.2), documented only in the [issue linked above](https://cs.android.com/search?q=macrobenchmark%2Fbuild.gradle&ss=androidx%2Fplatform%2Fframeworks%2Fsupport). Note that if you're looking for other examples, you can see usage [within the androidx repo](https://cs.android.com/search?q=macrobenchmark%2Fbuild.gradle&ss=androidx%2Fplatform%2Fframeworks%2Fsupport) (we haven't finished migrating to the new structure).

2) `com.android.test` (like in this project's sample)
Must use `self-instrumenting` property, so requires AGP 7.0.0-beta04 (It may work with older 7.0 betas, not sure which is the first version)

Apologies this isn't more straightforward - our intent was only to support 7.0-beta+, but happy to (best effort) try and help you make it work for older versions, especially if you don't care about local profiling experience.

**Comment by spinEdamageD** [comment]
Oh thanks a lot man, that looks great, I'll give it a try setting this up as a library module. Also I had another question, in regards to the requirement to have `<profileable android:shell="true" />` defined in the manifest, is that something that can be controlled via gradle so basically I can enable it only when I'm building for testing? 

The reason I ask is because it seems a bit of a security issue to have that always enabled in the manifest when we only need this for running these tests, so I'd rather not have it always there, and also, is that required if I only care about the single json results file? If this is the case then perhaps I could only add it when I need to build to debug a particular run or trace but leave it out when I'm just running the regular set of tests during CI or local testing.

Thanks again!

**Comment by ChrisCraik** [comment]
Unfortunately it's only controllable by the manifest currently - filed a bug for that here: https://issuetracker.google.com/194529081

The way that the profileable flag works is it only allows the shell process to profile your app, not any process on the system. Additionally, the main change it performs - enabling `android.os.Trace` APIs (which is necessary to get useful trace sections from within your process, and make them useful) is actually on by default in Android 12 ([1](https://android-review.googlesource.com/c/platform/frameworks/base/+/1587412/) [2](https://android-review.googlesource.com/c/platform/system/core/+/1581672/)). The flag will (eventually) not be required by the library starting with that OS version.

Few things you could try:
1. Use build variants to create a variant of your app with that manifest flag.
1. Enable profileable for all builds. This also enables you to directly profile your release build (with some upcoming changes to the Studio profilers, which will greatly increase accuracy)
1. Suppress the profileable warning and call [setAppTracingAllowed](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/os/Trace.java;l=177;drc=1241c9b6f20d0580198354674a82f5ca37cb56b6) from your target app via reflection conditionally. Who needs manifest flags anyway?

**Comment by spinEdamageD** [comment]
Oh thanks again for the response and for creating the bug.

As for your suggestions, for #1, I do currently have multiple variants, and yeah this is what I ideally want to do, but you mentioned that this is only controlled via the manifest, so how could I do this with build variants?

**Comment by ChrisCraik** [comment]
I haven't tried it but the docs mention a per-variant or per-build-type source set that gradle uses automatically: https://developer.android.com/studio/build/build-variants#sourceset-build

So in your app, alongside `src/main/AndroidManifest.xml` you could have `src/benchmark/AndroidManifest.xml` with just the one profileable tag - that should get enabled only for the benchmark variant by manifest merging (see lower in that page `Manifests are merged together into a single manifest. ...`).

**Comment by spinEdamageD** [comment]
Oh got it. Thanks again man. Will give that a try and hopefully that works.

**Comment by spinEdamageD** [comment]
Hey man, just wanted to let you know that I was able to make it work when I updated according to the docs using AGP v4, and well now that AGP v7 was just released, hopefully soon I can also update that to use the latest docs.

Having said that, I had a couple more questions that I was hoping you could please help me with or at least point me in the right direction.

1. I was curious exactly how the startup macrobenchmark test worked in terms of exactly when it marked the cold start as "complete". I've seen that usually they look at when the main activity is displayed in screen, but if a call to the `reportFullyDrawn` method was in the code, then that's the point where the cold start was marked as complete. I was hoping that this was also the case for this test, but just wanted to confirm.

2. And secondly, and slightly related to the above. I was wondering if it was possible to set up some sort of 'preparation' code before actually starting the full cold start test. The use case that I want to cover is that I actually need to log in the user before actually being able to properly measure cold start. I've seen that there is a 'setupBlock' in the call to `measureRepeated` in the benchmark rule, but I'm not sure if that runs in the same process as the test does every time or not, because in the particular cold start scenario I actually need to run this code once, and then afterwards starting the process of the test by killing and restarting the app process and I'm not sure this is how it works.

If you have any insights into the above I would really appreciate your help on this.

Thanks again!

**Comment by ChrisCraik** [comment]
1. reportFullyDrawn [support just landed this week](https://android-review.googlesource.com/c/platform/frameworks/support/+/1776807), so you can try it out with a [snapshot build now](https://androidx.dev/), or wait for the next alpha release, which should contain that change. It's a separate metric reported by StartupTimingMetric, but you'll need to make sure you wait at the end of the test (e.g. using UiAutomator to await for some component to display), so it doesn't complete measurements before your app is done

1. The setupBlock runs *after* the kill process (see [here](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:benchmark/benchmark-macro/src/main/java/androidx/benchmark/macro/Macrobenchmark.kt;l=276;drc=f40c94bb7b412071f3f47fac61b1db98e3bb74c6)) that ensures that your startup is cold. You can workaround this by simply calling scope.killProcess again at the end of your setupBlock, but that's isn't especially elegant. Is this login something you'd want to run at the very beginning of a test class? If so, it may make sense to use an `@BeforeClass` setup step to launch the package yourself, setup login, and then run all of your benchmarks, so you only have to login once. Would be nice to avoid logging in once per benchmark iteration, and setupBlock does just that.

**Comment by spinEdamageD** [comment]
Hey man, thanks for the quick reply. 

Sounds good, I'll keep an eye out for the next release and try using the snapshot for now. Is there a place where I can easily keep track of the releases? Both alpha and not would be great. And based on what you mention, if I understand correctly, it's possible for the measurement to complete before I actually call `reportFullyDrawn` so I do indeed need to setup the UiAutomator to wait for a component either way and it's not enough to just call that method. Would you by any chance happen to have an example of this?

As for the 2nd point, since we actually persist the login information between sessions, it would actually be enough to run the login only once before running all tests, but yeah this being a cold start test I do need this to run before the process is killed so that we measure true cold startup when the user is previously logged in. If I understand correctly, using a `@BeforeClass` should do the trick for this. Are there any docs on how that works or an example I can look at for that?

Thanks again!

**Comment by ChrisCraik** [comment]
Release notes are here, always updated with each release (alpha/beta/stable): https://developer.android.com/jetpack/androidx/releases/benchmark

Measurement completes when the block does, so if you call `startActivityAndWait()`, that waits for the activity to finish drawing its first frame, but doesn't wait for its content to be completely ready. You can do that waiting yourself with UI automator, e.g if you have some view type that only shows up when your content is ready (e.g. some list item that only is added to the hierarchy once your network or DB load completes), you can do:

```device.wait(Until.findObject(By.clazz("mypackage.SomeView")), /*timeoutMs*/ 3000)```

[`By`](https://developer.android.com/reference/androidx/test/uiautomator/By) has other ways to select what content to wait for, like an content description or textview content. If all else fails, you can always conservatively Thread.sleep(5000) or something, but that can be a wasteful if the time isn't accurate.

`@BeforeClass` is a [standard JUnit4 API](https://junit.org/junit4/javadoc/4.13/org/junit/BeforeClass.html), while [this](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:camera/integration-tests/uiwidgetstestapp/src/androidTest/java/androidx/camera/integration/uiwidgets/rotations/ImageAnalysisBaseTest.kt;l=135;drc=5c0d3b8d8475c8c55e7f29e8d9e430e7a918ebb6) is a random kotlin usage from within AndroidX codebase (useful to have a Kotlin example, since it needs companion class + @JvmStatic). It's just an annotation that makes a method run once, before any other code in the class is run.

**Comment by spinEdamageD** [comment]
Oh great. Thanks a lot for the info man. Will try that out. Have a nice weekend!

**Comment by spinEdamageD** [comment]
Hey man.. hope you're doing well. I was working on some other stuff and just got back to this and I've run into another issue which I was hoping you could please help me sort out.

So I'm trying to do the preparation for the test by logging the user beforehand, but now that I'm actually trying to set it up, I'm wondering how exactly this should work, because now that the test process is separate from the main app one, I can't think of an easy way to actually start the app and log the user in from the `@BeforeClass` code because at that point in time the main app code is not even running, so I'm just wondering if there's a way to set it up so that I can actually access the application object from the main app because I actually need this to log the user in.

I don't know if that makes sense or maybe I'm missing something, but the fact that we are now working with 2 separate processes has gotten me a bit confused as to how this would work, but I'm guessing this is a common requirement so I would think there's an easy way to do this.

Thanks again!

**Comment by ChrisCraik** [comment]
Hey, doing fine, hope you're doing fine too :D

> because now that the test process is separate from the main app one, I can't think of an easy way to actually start the app and log the user in from the @BeforeClass code because at that point in time the main app code is not even running

Apologies for not providing login samples - we definitely want to, but we're prioritizing older device support and metric improvements in macrobenchmark first. This is a good reminder for us to get to this at some point.

Two options to do this:

### UiAutomator

If you're trying to log the user in, and the app / activity isn't started - start it with UiAutomator, the same APIs used to drive the UI within a macrobenchmark. If you want to log in the user, use UiAutomator to enter test account username/password info into the login screen. Yes, benchmarks will kill the process in between tests, but they don't clear data (unless you do that) - the user will stay logged in.

See the [uiautomator sample](https://github.com/android/testing-samples/blob/main/ui/uiautomator/BasicSample), particularly the [sample test content](https://github.com/android/testing-samples/blob/cb136a96872d05330e08da6156a6acb4f61a1dde/ui/uiautomator/BasicSample/app/src/androidTest/java/com/example/android/testing/uiautomator/BasicSample/ChangeTextBehaviorTest.java#L59). Using UiAutomator to drive a separate test is certainly more cumbersome than more modern testing APIs, but it lets the test be completely independent from the app, without changing it.

### Custom entrypoint

There's another way as well though, if you want to avoid baking unrelated login UI into ui automator code - create a custom benchmark-only entrypoint that does this for you. Write it in application code, e.g. a `src/benchmark/BenchmarkActivity.kt`, and use that to set up (or clear) any global state you want in `Activity#onCreate`. Similarly, register it in `src/benchmark/AndroidManifest.xml`.

You could alternately use a [BroadcastReceiver](https://developer.android.com/reference/kotlin/android/content/BroadcastReceiver), and trigger that instead, but be sure to trigger it explicitly. To do that, specify the receiver class name, as the app may be force-stopped from a previous operation: `UiDevice.executeShellCommand("am broadcast -a $action -p $packageName/$receiverClassName")`. We do this [internally in benchmark](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:benchmark/benchmark-macro/src/main/java/androidx/benchmark/macro/CompilationMode.kt;l=110;drc=074c901bb93427d956ef32758cf7af6ddffb3397) for triggering a certain type of compilation.

This approach has the downside of changing app code, but is faster to get up and running, faster to run, and more stable (no mucking with timeouts).

**Comment by spinEdamageD** [comment]
Hey man, thanks for the quick response and for the suggestions, I'll give those options a try and hopefully I'm able to make it work without much hassle.

Another thing I was curious about is, I already have a way of essentially 'forcing' the login without the need to have the actual login screen showing in the UI, and I really only need access to the application object (so I basically only need to have the app running and have access to its main application object), so I was just wondering, would it be possible for me to start the app using UIAutomator and then just access its app object somehow and do this? or perhaps any other way?

If that's not possible then I'll definitely go with one of the two options suggested, but I just want to make sure I understand exactly how this works and see if I can find a way to do this with the code I already have and without having to add any more changes.

Thanks again!

**Comment by ChrisCraik** [comment]
Unfortunately, being in a separate process means you can't call any code in the other process without some inter-process means of communication (IPC).

You could use binder, activities, broadcasts, or content providers, but can't call into that code directly for the same reasons some other app on your phone can't call your app's code. It's completely separate memory spaces, with very different lifecycles.

**Comment by spinEdamageD** [comment]
Oh got it. Thanks again for your help on this and your quick reply. That makes sense. So I'll try using the UIAutomator as seen in the demo and just make sure to do the login process that way before starting the test.

**Comment by spinEdamageD** [comment]
Hey man, sorry to keep bugging you but I just encountered a new issue after updating to alpha07 version and I don't recall seeing any requirement to update the 'compileSdkVersion' anywhere so I'm just wondering if I missed something or what because from the demo I recall seeing that this was set to '30' and this is actually the value I have setup throughout so I don't think I can upgrade this in all my modules, but this is the error I'm getting now:

`
The minCompileSdk (31) specified in a dependency's AAR metadata (META-INF/com/android/build/gradle/aar-metadata.properties) is greater than this module's compileSdkVersion (android-30). Dependency: androidx.benchmark:benchmark-macro-junit4:1.1.0-alpha07.AAR metadata file: /root/.gradle/caches/transforms-3/ce1205dfc151790c2ea9a762ed59c613/transformed/jetified-benchmark-macro-junit4-1.1.0-alpha07/META-INF/com/android/build/gradle/aar-metadata.properties.
`

Thanks again!

**Comment by ChrisCraik** [comment]
**Compile** SDK is just the version of the SDK tools you're using at compile time - you can update it without having to handle platform behavior changes.

This compile sdk is bumped periodically across all non-stable AndroidX library versions any time we add support for a new platform version, but it should be easy to change on your side. It's considered trivial enough to not be included in patch notes, since it should be as simple as changing a number (and downloading the new SDK tools).

Note that it's different from **target** SDK, which often changes platform behavior, and requires you to change your app / behavior.

More info on the difference: https://medium.com/androiddevelopers/picking-your-compilesdkversion-minsdkversion-targetsdkversion-a098a0341ebd

**Comment by spinEdamageD** [comment]
Oh got it, thanks man. And do I need to update all my modules or with only the 'macroTests' one be enough?

**Comment by ChrisCraik** [comment]
I don't know, I would always just update it everywhere since it's easy.

**Comment by spinEdamageD** [comment]
Sounds good, thanks a lot man.

**Comment by spinEdamageD** [comment]
Hey @yolapop , sorry for the late reply, but I've been in vacation for the last two weeks and just got back. I was indeed able to make the test run, although I had to use the older guide because I'm using an AGP version older than 7. 

The guide is used for this is:
https://android.googlesource.com/platform/frameworks/support/+/870f4590b5d3749d6275497c30a1520998372e42/benchmark/docs/macrobenchmark.md

Having said that, I'm still stuck in being able to successfully log in the user before being able to actually run the full test as I would like (I'm trying to use UI Automator and a @BeforeClass to do this, but I keep getting a crash and since there is no video recording of that part, it hasn't been easy to figure out), but I had to work on something else for a bit so haven't had time to actually complete this.

Hopefully we can get a sample of how to do this with more detail, but so far that's where I'm stuck. 

If anything changes I'll post an update here, but hopefully that guide will help you out if you need to set this up using AGP < 7.

**Comment by androideveloper** [comment]
Hi. I was finally able to test the latest version of the macrobenchmark library on our app! Have a few questions here:

1. According to the [docs](https://developer.android.com/studio/profile/macrobenchmark#access-trace-files) this command `$ ./gradlew :macrobenchmark:connectedCheck` should copy files from the device to the host machine that runs the tests, however it doesn’t (android.enableAdditionalTestOutput=true is set in gradle.properties) 

2. There is an argument for changing the path of storing traces and JSON `additionalTestOutputDir`. How to pass it to the macrobenchmark? 

3. How to wait until `Activity.reportFullyDrawn` is called to get metric `fullyDrawnMs`? The hacky way is adding Thread.sleep for a few seconds to make sure it appears there. Is there a better way to do it?

The setup I have: 
Android Studio Arctic Fox | 2020.3.1 Patch 3
AGP 7.0.3
Device Pixel 2 with Android 11

**Comment by androideveloper** [comment]
Maybe you could help @ChrisCraik ?

**Comment by spinEdamageD** [comment]
In regards to the point 3 you mention, seems like there is not an ideal way to do this unless you manually wait until a view you know will for sure be displayed until after you made the call to `reportFullyDrawn`. I do think that it would be great to have a method built in the library to call when you want to wait until that method gets called rather than having to manually implement it, as I do think this will be a heavy use use-case, so this could work as a complement to `startActivityAndWait` which just waits until the first frame from the activity is displayed, but just a suggestion.

**Comment by androideveloper** [comment]
Absolutely agree! Would be helpful! Because most of the performance issues are coming after first frame and before fully drawn happens.

**Comment by keyboardsurfer** [comment]
This discussion seems to be resolved and I'm closing this issue. If you encounter issues with the sample or library in the future, please open a new [issue](new).
